### PR TITLE
[Hold] Created method to add, edit, and remove contributor

### PR DESCRIPTION
**Pull request by PatrickEGorman**
# Purpose

Allow users to add/remove contributors to nodes in API, view contributor details, and edit permissions and bibliographic status.
Answers https://github.com/CenterForOpenScience/osf.io/issues/3045

# Changes

Moved contributor serializer and added contributor detail serializer.  
Created special permissions class to determine whether a user can add contributors.
Created another permission class for user to view, edit, and delete individual contributors.  
Appended an add contributor option to contributor list page with permission and visibility options.
Created a user detail page with edit (visibility and permissions) and remove function
Added numerous api-tests

# Side effects

Allows users to add, and remove node contributors as well as edit their permission and bibliographic statuses.

### Add Users
Users are added by admins via user id.  There are options for permissions and bibliographic status.  The defaults are 'read' and 'write' for permissions and True for bibliographic status.  The permissions give 3 options of increasing user control over node.   Whatever the user selects, all lower tier options are also added.

### Edit Users
An admin can change a users visibility if they are not the only visible contributor
A user can change they're own visibility status from true to false regardless of admin status providing they're the not the only visible contributor

An admin can edit user permissions if they are not eliminating the only admin
A user can change they're own permissions regardless of admin status from 'admin' to 'write' to 'read' but not the other way, providing that they're not the only admin.  The edit function uses the same permissions hierarchy structure as the add function.

### Delete Users
An admin can remove a user if:
 - The user is not the only admin
 - The user is not the only visible contributor
 - The user is removing a non-admin contributor
A user can remove themselves regardless of admin status providing that they're not the only visible contributor or admin contributor.

# JSON

### Changes:
 - Added permissions list field
 - Added edit contributor link field

### Original

```json
{
    "data": [
        {
            "id": "h2pve",
            "fullname": "yarr",
            "given_name": "yarr",
            "middle_name": "",
            "family_name": "",
            "suffix": "",
            "date_registered": "2015-07-12T15:15:36.460000",
            "gravatar_url": "https://secure.gravatar.com/avatar/9d4b9fe9fbb545875c27d974ee5d7523?d=identicon&size=40",
            "employment_institutions": [],
            "educational_institutions": [],
            "social_accounts": {},
            "links": {
                "nodes": {
                    "relation": "http://localhost:8000/v2/users/h2pve/nodes/"
                },
                "html": "http://localhost:5000/h2pve/",
                "self": "http://localhost:8000/v2/users/h2pve/"
            },
            "bibliographic": true,
            "type": "users"
        },
        {
            "id": "zh6ji",
            "fullname": "test",
            "given_name": "test",
            "middle_name": "",
            "family_name": "",
            "suffix": "",
            "date_registered": "2015-06-10T13:31:39.244000",
            "gravatar_url": "https://secure.gravatar.com/avatar/dd46a756faad4727fb679320751f6dea?d=identicon&size=40",
            "employment_institutions": [],
            "educational_institutions": [],
            "social_accounts": {},
            "links": {
                "nodes": {
                    "relation": "http://localhost:8000/v2/users/zh6ji/nodes/"
                },
                "html": "http://localhost:5000/zh6ji/",
                "self": "http://localhost:8000/v2/users/zh6ji/"
            },
            "bibliographic": true,
            "type": "users"
        }
    ],
    "links": {
        "first": null,
        "last": null,
        "prev": null,
        "next": null,
        "meta": {
            "total": 2,
            "per_page": 10
        }
    }
}
```

### Updated
```json
{
    "data": [
        {
            "id": "h2pve",
            "fullname": "yarr",
            "given_name": "yarr",
            "middle_name": "",
            "family_name": "",
            "suffix": "",
            "date_registered": "2015-07-12T15:15:36.460000",
            "gravatar_url": "https://secure.gravatar.com/avatar/9d4b9fe9fbb545875c27d974ee5d7523?d=identicon&size=40",
            "employment_institutions": [],
            "educational_institutions": [],
            "social_accounts": {},
            "links": {
                "contributor-self": "http://localhost:8000/v2/nodes/emdk3/contributors/h2pve/",
                "html": "http://localhost:5000/h2pve/",
                "nodes": {
                    "relation": "http://localhost:8000/v2/users/h2pve/nodes/"
                },
                "self": "http://localhost:8000/v2/users/h2pve/"
            },
            "permissions": [
                "read",
                "write",
                "admin"
            ],
            "bibliographic": true,
            "type": "users"
        },
        {
            "id": "zh6ji",
            "fullname": "test",
            "given_name": "test",
            "middle_name": "",
            "family_name": "",
            "suffix": "",
            "date_registered": "2015-06-10T13:31:39.244000",
            "gravatar_url": "https://secure.gravatar.com/avatar/dd46a756faad4727fb679320751f6dea?d=identicon&size=40",
            "employment_institutions": [],
            "educational_institutions": [],
            "social_accounts": {},
            "links": {
                "contributor-self": "http://localhost:8000/v2/nodes/emdk3/contributors/zh6ji/",
                "html": "http://localhost:5000/zh6ji/",
                "nodes": {
                    "relation": "http://localhost:8000/v2/users/zh6ji/nodes/"
                },
                "self": "http://localhost:8000/v2/users/zh6ji/"
            },
            "permissions": [
                "read",
                "write",
                "admin"
            ],
            "bibliographic": true,
            "type": "users"
        }
    ],
    "links": {
        "first": null,
        "last": null,
        "prev": null,
        "next": null,
        "meta": {
            "total": 2,
            "per_page": 10
        }
    }
}
```

## (New) Contributor Detail Page:
```json
{
    "data": {
        "id": "c2d7f",
        "fullname": "Freddie Mercury12",
        "given_name": "Freddie",
        "middle_name": "",
        "family_name": "Mercury12",
        "suffix": "",
        "date_registered": "2015-06-09T20:23:37.605000",
        "gravatar_url": "https://secure.gravatar.com/avatar/21ff3beadebb3d768b083e0a80e4d98d?d=identicon&size=40",
        "employment_institutions": [],
        "educational_institutions": [],
        "social_accounts": {},
        "links": {
            "contributor-self": "http://localhost:8000/v2/nodes/rwhq7/contributors/c2d7f/",
            "html": "http://localhost:5000/c2d7f/",
            "nodes": {
                "relation": "http://localhost:8000/v2/users/c2d7f/nodes/"
            },
            "self": "http://localhost:8000/v2/users/c2d7f/"
        },
        "permissions": [
            "read",
            "write",
            "admin"
        ],
        "bibliographic": false,
        "type": "users"
    }
}
```

### CODE REVIEW

**Comment by lbanner** [comment]
Nice "To Do" list @PatrickEGorman ! That's a good practice. Thanks.

**Comment by PatrickEGorman** [comment]
Thanks.  I copied it off of somebody else because it seemed like a good idea

**Comment by PatrickEGorman** [comment]
My apologies for running too many Travis CI tests.  I now know to use [skip ci] in future pull requests

File: api/base/settings/defaults.py [hunk 1]
```diff
@@ -79,7 +79,7 @@
     # Needs to go before CommonMiddleware, so that transactions are always started,
     # even in the event of a redirect. CommonMiddleware may cause other middlewares'
     # process_request to be skipped, e.g. when a trailing slash is omitted
-    'api.base.middleware.DjangoGlobalMiddleware',
+    #'api.base.middleware.DjangoGlobalMiddleware',
```
Comment (brianjgeiger) [inline api/base/settings/defaults.py]: Why is this commented out?

File: api/nodes/permissions.py [hunk 2]
```diff
@@ -22,6 +22,50 @@ def has_object_permission(self, request, view, obj):
...
+        else:
+            node = Node.load(request.parser_context['kwargs']['node_id'])
+        auth = get_user_auth(request)
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        else:
+            return node.has_permission(auth.user, 'admin')
+
+
+class ContributorPermissions(permissions.BasePermission):
```
Comment (brianjgeiger) [inline api/nodes/permissions.py]: This should be renamed. `Contributor` would probably work best with the existing permissions.

File: api/nodes/serializers.py [hunk 3]
```diff
@@ -127,6 +128,114 @@ def update(self, instance, validated_data):
         return instance
 
 
+class ContributorSerializer(UserSerializer):
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: This should be NodeContributorsSerializer

File: api/nodes/serializers.py [hunk 4]
```diff
@@ -127,6 +128,114 @@ def update(self, instance, validated_data):
...
+            raise NotFound('User with id {} cannot be found.'.format(validated_data['_id']))
+        elif contributor in node.contributors:
+            raise ValidationError('User {} already is a contributor.'.format(contributor.username))
+        else:
+            node.add_contributor(contributor=contributor, auth=auth, save=True)
+            contributor.node_id = node._id
+        return contributor
+
+
+class ContributorDetailSerializer(ContributorSerializer):
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: How is this different from NodeContributorsSerializer (currently ContributorSerializer)? I think you should be able to handle everything with single serializer.

File: api/nodes/views.py [hunk 5]
```diff
@@ -142,12 +142,63 @@ def get_default_queryset(self):
...
+        user.bibliographic = node.get_visible(user)
+        return user
+
+    # overrides DestroyAPIView
+    def perform_destroy(self, instance):
+        node = self.get_node()
+        current_user = self.request.user
+        auth = Auth(current_user)
+        if not self.has_multiple_admins(node) and node.has_permission(instance, 'admin'):
+            raise PermissionDenied("Must have at least one registered admin contributor")
```
Comment (brianjgeiger) [inline api/nodes/views.py]: Hmmm. This might should be a bad request rather than permission denied.

File: api/nodes/views.py [hunk 6]
```diff
@@ -142,12 +142,63 @@ def get_default_queryset(self):
...
+        if node.get_visible(instance) and len(node.visible_contributors) == 1:
+            raise PermissionDenied("Must have at least one visible contributor")
```
Comment (brianjgeiger) [inline api/nodes/views.py]: Same here. It's not that the contributor doesn't have the proper permissions to make the change, it's that the request is improper For Reasons.

**Comment by brianjgeiger** [comment]
I still have to go through the tests, but consider this a first pass complete. :penguin: There are a couple of big things to fix, and I'll do a more thorough check once you've taken care of those.

File: api/base/settings/defaults.py (hunk 1 above)
Comment (PatrickEGorman) [inline api/base/settings/defaults.py]: I was getting strange errors and it wasn't processing exceptions properly.  I noted in my pull request that I can comment it out and push again if you want.

File: api/nodes/serializers.py (hunk 4 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: That makes sense.  I'll go to work on combining and refactoring the serializers

File: api/nodes/views.py (hunk 5 above)
Comment (PatrickEGorman) [inline api/nodes/views.py]: Good catch.  I used bad request for the same thing in other places so this would be inconsistent as well.

File: api/base/settings/defaults.py (hunk 1 above)
Comment (brianjgeiger) [inline api/base/settings/defaults.py]: @abought Is this related at all to what you've been up to?

**Comment by PatrickEGorman** [comment]
Sounds good.  I'll try to get some json samples up here and screenshot the docs today, then go to work on this stuff/ look for other issues tomorrow.

File: api/base/settings/defaults.py (hunk 1 above)
Comment (brianjgeiger) [inline api/base/settings/defaults.py]: @PatrickEGorman Uncomment that line and let's see what happens. Thanks!

Comment (abought) [inline api/base/settings/defaults.py]: There were previously errors with something else- people were commenting out `FlaskRequestMiddleware` until the `DjangoGlobalsMiddleware` fix landed in #3389. (one of the symptoms of the old bug was weird exception behavior)

If the new `DjangoGlobalsMiddleware` is also having problems, then absolutely let me know and send a test case. But please *don't* commit workarounds to git without at least reporting the bug. :camel:

Comment (PatrickEGorman) [inline api/base/settings/defaults.py]: Sounds good.  I'll make sure to ping you if you have any problems next time

**Comment by brianjgeiger** [comment]
The tests that I see failing are failing incorrectly. If you're doing a request that you are expecting to return with something that's in the 400 or 500 range, you have to tell it `expect_errors=True` in the request. I didn't check the later ones, but they may be something similar.

**Comment by brianjgeiger** [comment]
I may have misread, so I'm looking a bit closer.

**Comment by brianjgeiger** [comment]
Hmmmm. Okay, that's odd.

**Comment by PatrickEGorman** [comment]
It appears that by uncommenting the line, a lot of my previously passing tests fail.  I notice, however that only my tests are failing.  I'll have to look into this more

File: api/nodes/serializers.py (hunk 4 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: I created this serializer to add a permission choicefield and make the boolean field not write only.  Is it possible to make a field usable in a specific page?  Do you want me to add these options to the add user option in the userlist page?

Comment (brianjgeiger) [inline api/nodes/serializers.py]: From an API standpoint, the information should be available everywhere. The UI can determine if it wants to display the information, but we don't want specific API endpoints that are slightly different here and there.

Comment (PatrickEGorman) [inline api/nodes/serializers.py]: Okay.  I'll work on that.  Thanks for the help

File: api/nodes/serializers.py [hunk 7]
```diff
@@ -127,6 +128,133 @@ def update(self, instance, validated_data):
...
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+
+    id = ser.CharField(read_only=True, source='_id')
+
+    # Overridden to allow blank for user to not change status
+    permission = ser.ChoiceField(choices=['admin', 'read', 'write'], write_only=True, allow_blank=True)
+
+    # Overridden to remove contributor detail link
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: I think it's pretty standard for detail items to have links to themselves. Verify by checking out the node list and node detail. There should be a self link in both of those. If so, go ahead and take out this override.

Comment (PatrickEGorman) [inline api/nodes/serializers.py]: Indeed it does.  Will make change

**Comment by abought** [comment]
> It appears that by uncommenting the line, a lot of my previously passing tests fail. I notice, however that only my tests are failing. I'll have to look into this more

Documenting discussion with @PatrickEGorman : some of these tests were failing because they changed the object in an API call, but did not fetch the newest data afterwards.

Without DjangoGlobalsMiddleware, they could pass because the API call and the unit test were sharing the same in-memory database cache. So there was no need to get the newest data after the "external" API call was done... but only because, like many old tests, it was relying on a bug.

@PatrickEGorman will use `database_object.reload()` where appropriate and report back on any remaining test failures.

**Comment by brianjgeiger** [comment]
It might be worth doing a GET on the object and comparing that with the expected value first, then doing the reload() and verifying it's updated in the database, as part of each of these tests.

**Comment by PatrickEGorman** [comment]
Spent some time looking things over.  My tests were not looking so good earlier but I fixed them up a lot.  Ready second pass when you are.

**Comment by brianjgeiger** [comment]
ping @lyndsysimon for review. Medium priority.

File: api/nodes/serializers.py [hunk 8]
```diff
@@ -130,6 +131,125 @@ def update(self, instance, validated_data):
...
+
+    id = ser.CharField(source='_id')
+    permissions = ser.ListField(read_only=True)
+    bibliographic = ser.BooleanField(initial=True, help_text='Whether the user will be included in citations for '
+                                                               'this node or not')
+    permission = ser.ChoiceField(choices=['read', 'write', 'admin'], initial='write', write_only=True)
+    local_filterable = frozenset(['permission', 'bibliographic'])
+    filterable_fields = frozenset.union(UserSerializer.filterable_fields, local_filterable)
+
+    fullname = ser.CharField(read_only=True, help_text='Display name used in the general user interface')
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: Aren't all these in the UserSerializer? I don't think you should need to redefine them.

File: api/nodes/views.py [hunk 9]
```diff
@@ -142,12 +142,63 @@ def get_default_queryset(self):
...
+        node = self.get_node()
+        current_user = self.request.user
+        auth = Auth(current_user)
+        if not self.has_multiple_admins(node) and node.has_permission(instance, 'admin'):
+            raise ValidationError("Must have at least one registered admin contributor")
+        if len(node.visible_contributors) == 1 and node.get_visible(instance):
+            raise ValidationError("Must have at least one visible contributor")
+        node.remove_contributor(instance, auth)
+
+    def has_multiple_admins(self, node):
```
Comment (brianjgeiger) [inline api/nodes/views.py]: You shouldn't need this in both the view and the serializer. In a pinch, you could define in one place and import, but I expect you could just pass the results from the view to the serializer, yes?

File: api/nodes/serializers.py [hunk 10]
```diff
@@ -130,6 +131,125 @@ def update(self, instance, validated_data):
...
+    })
+
+    def absolute_url(self, obj):
+        return obj.absolute_url
+
+    def create(self, validated_data):
+        current_user = self.context['request'].user
+        auth = Auth(current_user)
+        node = self.context['view'].get_node()
+        user = User.load(validated_data['_id'])
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: You might be able to use get_object_or_404() here, then check to make sure the user isn't in contributors after.

File: api/nodes/serializers.py [hunk 11]
```diff
@@ -130,6 +131,125 @@ def update(self, instance, validated_data):
...
+        node.add_contributor(contributor=user, auth=auth, save=True)
+        self.set_visibility(bibliographic, user, node)
+        self.set_permissions(permission_field, user, node)
+        user.node_id = node._id
+        user.bibliographic = node.get_visible(user)
+        return user
+
+    def set_visibility(self, bibliographic, user, node, is_admin=True, is_current=False):
+        if is_admin and bibliographic:
+            node.set_visible(user, True, save=True)
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: Does this raise errors if you try to do things inappropriately? You might be able to save yourself some checking here by handling errors raised rather than checking ahead of time. We are a bit inconsistent about where we handle these things, though, so if it's not on the model, then definitely keep these checks.

File: api/nodes/serializers.py (hunk 8 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: The difference is that they're read only.  Is there a way to make them read only without redefining them?

Comment (brianjgeiger) [inline api/nodes/serializers.py]: Yeah, we probably had this discussion before. You *might* be able to iterate through the field definitions and set read_only to true. If you could do that in a generic way, it would be good. Another way would be to make the actual user a related field rather than duplicating the info. That way people could grab the user info if need be, or use your upcoming include functionality once that's ready.

File: api/nodes/views.py (hunk 9 above)
Comment (PatrickEGorman) [inline api/nodes/views.py]: Moved to utils.  It's used in two different views, so simply putting it in one wouldn't work

File: api/nodes/serializers.py (hunk 8 above)
Comment (brianjgeiger) [inline api/nodes/serializers.py]: Probably the second way with the related field is the "right" way to do it.

Comment (PatrickEGorman) [inline api/nodes/serializers.py]: Okay. I like the idea of a relation for now.  When I get this include request working, I think it'd be easier to convert over

File: api/nodes/serializers.py (hunk 11 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: The model returns false if the user is in node.contributors but does not check to see if a user exists...

Comment (PatrickEGorman) [inline api/nodes/serializers.py]: Actually it does raise an attribute error when trying to get user attributes
```
try:
            added = node.add_contributor(contributor=user, auth=auth, save=True)
            if not added:
                raise ValidationError('User {} already is a contributor.'.format(user.username))
        except AttributeError:
            raise NotFound('User with id {} cannot be found.'.format(validated_data['_id']))
```
Works fine if you think that's better

File: api/nodes/serializers.py (hunk 10 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: I'm confused.  Did you mean to put this after the remove contributor method?

Comment (brianjgeiger) [inline api/nodes/serializers.py]: No. get_object_or_404 will load an object from the ID or from a query, or it will 404 if it doesn't exist. Which is what these three lines do.

File: api/nodes/serializers.py (hunk 11 above)
Comment (brianjgeiger) [inline api/nodes/serializers.py]: Yeah, that way if it changes, it only has to change on the model and not in both places.

Comment (brianjgeiger) [inline api/nodes/serializers.py]: And chances are you don't have to do anything to use it. See if the error bubbles up properly if you just try to set the visibility under the improper circumstances.

File: api/nodes/serializers.py (hunk 8 above)
Comment (PatrickEGorman) [inline api/nodes/serializers.py]: So I came up with a simple fix of just making it extend jsonapi serializer rather than the user serializer and setting all the fields separately.   Thoughts?

**Comment by PatrickEGorman** [comment]
Closed to create new branch off of develop
